#!/usr/bin/env python3
# -*- encoding: utf-8 -*-

import os
import click
import subprocess
from explore_pipolin.utilities import CONTEXT_SETTINGS
from Bio import SeqIO
from io import StringIO


def download_genomes_ncbi(metadata_file, out_dir):
    os.makedirs(out_dir, exist_ok=True)
    with open(metadata_file) as inf:
        for num, line in enumerate(inf):
            print(f'Downloading the {num + 1} genome...')
            assembly_id, _, _, ids_string = line.strip().split('\t')

            if ids_string != '-':
                seqs = []
                seqs_batch_request = subprocess.run(['epost', '-id', ids_string, '-db', 'nucleotide'],
                                                    stdout=subprocess.PIPE)
                seqs_batch_fetched = subprocess.run(['efetch', '-format', 'fasta'], input=seqs_batch_request.stdout,
                                                    stdout=subprocess.PIPE)
                seqs.extend(SeqIO.parse(StringIO(seqs_batch_fetched.stdout.decode(encoding='UTF8')), format='fasta'))

                if len(ids_string.split(sep=',')) != len(seqs):
                    raise AssertionError('The number of downloaded sequences (contigs) is wrong!!!')

                with open(os.path.join(out_dir, assembly_id + '.fa'), 'w') as ouf:
                    SeqIO.write(seqs, ouf, format='fasta')
            else:
                print('The genome is not in NCBI! PASS!')


@click.command(context_settings=CONTEXT_SETTINGS)
@click.argument('metadata-file', type=click.Path(exists=True))
@click.argument('out-dir', type=click.Path())
def main(metadata_file, out_dir):
    """
    Given METADATA_FILE, generated by download_metadata_ncbi.py, the genome sequences in
    FASTA format will be downloaded from NCBI to the OUT_DIR.
    NOTE: requires "ncbi-entrez-direct" package!
    """
    download_genomes_ncbi(metadata_file, out_dir)


if __name__ == '__main__':
    main()
