import os
from typing import List, Tuple

import click
import subprocess
from Bio import SeqIO
from io import StringIO
from explore_pipolin.common import CONTEXT_SETTINGS
from explore_pipolin.utilities.external_tools import subprocess_with_retries


def read_metadata_file(metadata_file: str) -> List:
    with open(metadata_file) as inf:
        inf.readline()   # skip header line
        lines = []

        for line in inf:
            if len(line.strip().split('\t')) == 6:
                lines.append(line.strip())

    return lines


def extract_data(metadata_file_lines: List) -> Tuple[List, List, List, List]:
    assembly_accs = [line.split('\t')[1] for line in metadata_file_lines]
    strain_names = ['_'.join(line.split('\t')[3].split(' ')) for line in metadata_file_lines]
    is_plasmid = [line.split('\t')[4] for line in metadata_file_lines]
    gb_ids = [line.split('\t')[5] for line in metadata_file_lines]
    return assembly_accs, strain_names, is_plasmid, gb_ids


def download_genome_seqs(gb_ids):
    seqs = []
    seqs_batch_request = subprocess_with_retries(['epost', '-id', gb_ids, '-db', 'nucleotide'],
                                                 input='', stdout=subprocess.PIPE)
    seqs_batch_fetched = subprocess_with_retries(['efetch', '-format', 'fasta'], input=seqs_batch_request.stdout,
                                                 stdout=subprocess.PIPE)
    seqs.extend(SeqIO.parse(StringIO(seqs_batch_fetched.stdout.decode(encoding='UTF8')), format='fasta'))

    if len(gb_ids.split(sep=',')) != len(seqs):
        raise AssertionError('The number of downloaded sequences (contigs) is wrong!!!')

    return seqs


@click.command(context_settings=CONTEXT_SETTINGS)
@click.argument('metadata-file', type=click.Path(exists=True))
@click.argument('out-dir', type=click.Path())
@click.option('--ncbi-api-key', default='')
def download_genomes_ncbi(metadata_file, out_dir, ncbi_api_key):
    """
    Given METADATA_FILE, generated by download_metadata_ncbi, the genome sequences in
    FASTA format will be downloaded from NCBI to the OUT_DIR.
    NOTE: requires "ncbi-entrez-direct" package!
    """
    os.environ['NCBI_API_KEY'] = ncbi_api_key

    os.makedirs(out_dir, exist_ok=True)
    plasmids_dir = os.path.join(out_dir, 'plasmids')
    os.makedirs(plasmids_dir, exist_ok=True)

    metadata_file_lines = read_metadata_file(metadata_file=metadata_file)
    assembly_accs, strains, is_plasmid_list, gb_ids = extract_data(metadata_file_lines=metadata_file_lines)

    for assembly_acc, strain, is_plasmid, gb in zip(assembly_accs, strains, is_plasmid_list, gb_ids):
        print(f'>>> Downloading sequences for the assembly {assembly_acc}')
        seqs = download_genome_seqs(gb_ids=gb)
        if is_plasmid == 'True':
            with open(os.path.join(plasmids_dir, assembly_acc + '_' + strain + '.fa'), 'w') as ouf:
                SeqIO.write(seqs, ouf, format='fasta')
        else:
            with open(os.path.join(out_dir, assembly_acc + '_' + strain + '.fa'), 'w') as ouf:
                SeqIO.write(seqs, ouf, format='fasta')


if __name__ == '__main__':
    download_genomes_ncbi()
